# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/ContentIterator.h
# Commit: 42c11ce3074f
# Full Hash: 42c11ce3074f90c1cfa4dd63bf0556cf986eb718
# Author: Sean Feng <sefeng@mozilla.com>
# Date: 2025-05-09 04:15:00
# Regressor Bug: 1932150
# File Overlap Count: 1
# Description:
#   Bug 1932150 - ContentSubtreeIterator #1: Make the end container related logic in ContentSubtreeIterator more robust r=jjaschke,smaug,dom-core
#   
#   This is more like a fix to the current logic. The end container could
#   be in a shadow tree of a shadow host, or in the light DOM of a shadow
#   host.
# ==============================================================================

diff -r 82ee04923c59 -r 42c11ce3074f dom/base/ContentIterator.h
--- a/dom/base/ContentIterator.h	Thu May 08 13:37:21 2025 +0000
+++ b/dom/base/ContentIterator.h	Thu May 08 13:37:21 2025 +0000
@@ -97,6 +97,20 @@
       nsIContent* aRoot,
       dom::AllowRangeCrossShadowBoundary aAllowCrossShadowBoundary);
 
+  struct AncestorInfo {
+    nsIContent* mAncestor = nullptr;
+    // mIsDescendantInShadowTree is used to determine if we should go
+    // dive into the shadow tree or regular light DOM tree if mAncestor
+    // is a shadow host. It should always be false otherwise.
+    bool mIsDescendantInShadowTree = false;
+  };
+
+  class InclusiveAncestorComparator {
+   public:
+    bool Equals(const AncestorInfo& aA, const nsINode* aB) const {
+      return aA.mAncestor == aB;
+    }
+  };
   // Get the next/previous sibling of aNode, or its parent's, or grandparent's,
   // etc.  Returns null if aNode and all its ancestors have no next/previous
   // sibling.
@@ -330,7 +344,7 @@
   RefPtr<dom::AbstractRange> mRange;
 
   // See <https://dom.spec.whatwg.org/#concept-tree-inclusive-ancestor>.
-  AutoTArray<nsIContent*, 8> mInclusiveAncestorsOfEndContainer;
+  AutoTArray<AncestorInfo, 8> mInclusiveAncestorsOfEndContainer;
 
   // Whether this iterator allows to iterate nodes across shadow boundary.
   dom::AllowRangeCrossShadowBoundary mAllowCrossShadowBoundary =
